### PR TITLE
feat(tofs): add support of TOFS

### PR DESCRIPTION
**Pull request by n-rodriguez**

### CODE REVIEW

**Comment by myii** [comment]
Initial feedback, I haven't looked at this in any detail at all.  But in terms of `yamllint`, this will cause an error with Jinja tags in the YAML files.  Ideally, we need to resolve this by pre or post-processing in `map.jinja` (or similar).  Worst case, we can ignore the YAML files containing the Jinja, but that then loses the benefit of having `yamllint`.

**Comment by n-rodriguez** [comment]
@myii  ping

**Comment by myii** [comment]
@n-rodriguez I'll check this soon, thanks for the reminder.

File: docs/TOFS_pattern.rst [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+           - salt://ntp/files/theminion/etc/ntp.conf_alt.jinja
+           - salt://ntp/files/theminion/etc/ntp.conf.jinja
+           - salt://ntp/files/Debian/etc/ntp.conf_alt.jinja
+           - salt://ntp/files/Debian/etc/ntp.conf.jinja
+           - salt://ntp/files/default/etc/ntp.conf_alt.jinja
+           - salt://ntp/files/default/etc/ntp.conf.jinja
+
+Note: This does *not* override the default value.
+Rather, the value from the pillar/config is prepended to the default.
+
```
Comment (myii) [inline docs/TOFS_pattern.rst]: Please get the updated version of this document from the `template-formula`.  You'll see a section at the end here:

* Using sub-directories for `components`

**Review by myii — CHANGES_REQUESTED** [review]
First review after checking with https://github.com/myii/ssf-formula/pull/20.

File: docs/TOFS_pattern.rst (hunk 1 above)
Comment (n-rodriguez) [reply docs/TOFS_pattern.rst]: done :)

**Comment by myii** [comment]
@n-rodriguez Can you check all of the Travis failures?  From the first couple, there's a common error:

```
       local:
           Data failed to compile:
       ----------
           Rendering SLS 'base:influxdb.package.install' failed: Jinja variable 'dict object' has no attribute 'toml_pkg'
```

**Comment by n-rodriguez** [comment]
@myii it seems that influxdb is not available on this OpenSuse version (leap15) : https://software.opensuse.org/package/influxdb

**Comment by n-rodriguez** [comment]
default-centos-6-2017-7-py2 doesn't have any `python-pytoml` package

**Comment by n-rodriguez** [comment]
`default-amazonlinux-2-2019-2-py2` doesn't have any `python-pytoml` package

**Comment by myii** [comment]
@n-rodriguez Like last time, do you want to enable all of the instances, to see which ones work?

**Comment by n-rodriguez** [comment]
> @n-rodriguez Like last time, do you want to enable all of the instances, to see which ones work?

done : https://travis-ci.com/saltstack-formulas/influxdb-formula/builds/126296726

**Comment by myii** [comment]
@n-rodriguez The `opensuse` instance was working prior to this (specifically [`default-opensuse-leap-15-2018-3-py2`](https://travis-ci.com/saltstack-formulas/influxdb-formula/jobs/234482027)).  Can we get that working again?

File: influxdb/package/install.sls [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
+# vim: ft=sls
+
+{#- Get the `tplroot` from `tpldir` #}
+{%- set tplroot = tpldir.split('/')[0] %}
+{%- from tplroot ~ "/map.jinja" import influxdb with context %}
+
+influxdb/package/install:
+  pkg.installed:
+    - name: {{ influxdb.pkg.name }}
```
Comment (aboe76) [inline influxdb/package/install.sls]: Maybe a fix for the opensuse failure, test with refresh in pkg.installed
```
influxdb/package/install:
  pkg.installed:
    - name: {{ influxdb.pkg.name }}
    - refresh: true
```

Comment (n-rodriguez) [reply influxdb/package/install.sls]: It seems that `influxdb` has been removed : 

```sh
6f14d870814f:/etc/zypp/repos.d # zypper search influx
Loading repository data...
Reading installed packages...

S | Name                                  | Summary                                                           | Type      
--+---------------------------------------+-------------------------------------------------------------------+-----------
  | golang-github-influxdata-usage-client | A Golang client for speaking to the Anonymous Usage Reporting API | package   
  | golang-github-influxdata-usage-client | A Golang client for speaking to the Anonymous Usage Reporting API | srcpackage
  | pcp-export-pcp2influxdb               | Performance Co-Pilot tools for exporting PCP metrics to InfluxDB  | package   
  | python2-influxdb                      | InfluxDB client                                                   | package   
  | python3-influxdb                      | InfluxDB client                                                   | package   ```

Comment (myii) [reply influxdb/package/install.sls]: @n-rodriguez Could we use one of the experimental or community packages instead?

* https://software.opensuse.org/package/influxdb

Comment (n-rodriguez) [reply influxdb/package/install.sls]: it works

**Review by aboe76 — APPROVED** [review]

**Comment by n-rodriguez** [comment]
@myii ping

**Comment by myii** [comment]
@n-rodriguez OK, to get an idea of what this PR is doing, I've taken advantage of the new `debug` mode to compare runs:

* https://travis-ci.org/myii/influxdb-formula/builds/590033879 (before)
* https://travis-ci.org/myii/influxdb-formula/builds/590038192 (after)

In terms of TOFS specifically, there wasn't much to check.  In fact, before and after aren't even the same states (when searching for `source:`):

https://travis-ci.org/myii/influxdb-formula/jobs/590033880#L1099-L1102 (before):

```salt
       influxdb_logrotate:
         file.managed:
           - name: /etc/logrotate.d/influxdb
           - source: salt://influxdb/files/logrotate.conf.jinja
```

https://travis-ci.org/myii/influxdb-formula/jobs/590038193#L991-L997 (after):

```salt
       influxdb/config/install:
         file.managed:
           - name: /etc/influxdb/influxdb.conf
           - source: 
             - salt://influxdb/files/1d1421e8f779/influxdb.tmpl.jinja
             - salt://influxdb/files/Debian/influxdb.tmpl.jinja
             - salt://influxdb/files/default/influxdb.tmpl.jinja
```

Just checked right now using the following search:

* https://github.com/saltstack-formulas/influxdb-formula/search?utf8=%E2%9C%93&q=file.managed&type=

So it looks like I'm going to have to do both runs again, with more states defined in `kitchen.yml`.  Let me get back with that feedback, when I get a chance to do that.

**Comment by myii** [comment]
@n-rodriguez So `influxdb.cli` has been removed?  Then this PR involves a breaking change (which isn't an issue but) that needs to be added to the commit message.

**Comment by myii** [comment]
@n-rodriguez OK, I only needed to modify the run done before TOFS, to get the necessary comparisons:

* https://travis-ci.org/myii/influxdb-formula/builds/590048078

This breakdown is part of the familiarisation, so if I ask questions, please don't take it the wrong way.

https://travis-ci.org/myii/influxdb-formula/jobs/590048079#L1047-L1050:

```salt
       influxdb_config:
         file.managed:
           - name: /etc/influxdb/config.toml
           - source: salt://influxdb/files/influxdb.config.jinja
```

* What's this replaced by?

https://travis-ci.org/myii/influxdb-formula/jobs/590048079#L1065-L1068:

```salt
       influxdb_default:
         file.managed:
           - name: /etc/default/influxdb
           - source: salt://influxdb/files/influxdb.etc_default
```

* What's this replaced by?

https://travis-ci.org/myii/influxdb-formula/jobs/590048079#L1162-L1165:

```salt
       influxdb_logrotate:
         file.managed:
           - name: /etc/logrotate.d/influxdb
           - source: salt://influxdb/files/logrotate.conf.jinja
```

* What's this replaced by?

---

If anything else is being removed by this PR, it will also need to be added as a breaking change.  Multiple breaking changes can be supplied in the commit body, like has been done here:

* https://github.com/saltstack-formulas/libvirt-formula/commit/8adfe9e

**Comment by sticky-note** [comment]
@myii @n-rodriguez Why did you closed this ??

**Pull request closed without merging**